tests/integration/test_types_integration.py

```diff
@@ -794,14 +794,92 @@ def test_array(trino_connection):
+
+    # temporal types
+    tz_india = create_timezone("+05:30")
+    tz_new_york = create_timezone("-05:00")
+    tz_los_angeles = create_timezone("America/Los_Angeles")
+    time_1 = time(1, 1, 1)
+    time_2 = time(23, 59, 59)
+    datetime_1 = datetime(1970, 1, 1, 1, 1, 1)
+    datetime_2 = datetime(2023, 1, 1, 23, 59, 59)
     SqlTest(trino_connection) \
-        .add_field(sql="CAST(null AS MAP(VARCHAR, INTEGER))", python=None) \
-        .add_field(sql="MAP(ARRAY['a', 'b'], ARRAY[1, null])", python={'a': 1, 'b': None}) \
-        .add_field(sql="MAP(ARRAY['a', 'b'], ARRAY[2.4, null])", python={'a': Decimal("2.4"), 'b': None}) \
-        .add_field(sql="MAP(ARRAY[2.4, 4.8], ARRAY[CAST(4.9E-324 AS DOUBLE), null])",
-                   python={Decimal("2.4"): 5e-324, Decimal("4.8"): None}) \
+        .add_field(sql="MAP(ARRAY[DATE '1970-01-01', DATE '2023-01-01'], ARRAY[DATE '1970-01-01', null])",
+                   python={date(1970, 1, 1): date(1970, 1, 1), date(2023, 1, 1): None}) \
+        .add_field(sql="MAP(ARRAY[TIME '01:01:01', TIME '23:59:59'], ARRAY[TIME '01:01:01', null])",
+                   python={time_1: time_1, time_2: None}) \
+        .add_field(sql="MAP("
+                       "ARRAY[TIME '01:01:01 +05:30', TIME '23:59:59 -05:00'], "
+                       "ARRAY[TIME '01:01:01 +05:30', null])",
+                   python={time_1.replace(tzinfo=tz_india): time_1.replace(tzinfo=tz_india),
+                           time_2.replace(tzinfo=tz_new_york): None}) \
+        .add_field(sql="MAP("
+                       "ARRAY[TIMESTAMP '1970-01-01 01:01:01', TIMESTAMP '2023-01-01 23:59:59'], "
+                       "ARRAY[TIMESTAMP '1970-01-01 01:01:01', null])",
+                   python={datetime_1: datetime_1, datetime_2: None}) \
+        .add_field(sql="MAP("
+                       "ARRAY[TIMESTAMP '1970-01-01 01:01:01 +05:30', TIMESTAMP '2023-01-01 23:59:59 America/Los_Angeles'], "  # noqa: E501
+                       "ARRAY[TIMESTAMP '1970-01-01 01:01:01 +05:30', null])",
+                   python={datetime_1.replace(tzinfo=tz_india): datetime_1.replace(tzinfo=tz_india),
+                           datetime_2.replace(tzinfo=tz_los_angeles): None}) \
         .execute()
 
+    # structural types - note that none of these below tests work in the Trino JDBC Driver either.
+    # Unhashable types like lists and dicts cannot be used as keys so these values cannot be represented as Python
+    #  objects at all.
+    # .add_field(sql="MAP(ARRAY[ARRAY[1, 2]], ARRAY[null])", python={[1, 2]: None})
+    # .add_field(sql="MAP(ARRAY[MAP(ARRAY[1], ARRAY[2])], ARRAY[null])", python={{1: 2}: None})
+
+    # TODO: fails because server sends [[{"[1, 2]":null}]] as response whereas it sends [[[1,2]]] as response for ROW
+    #  types that are not enclosed in a MAP while the RowValueMapper expects values to be lists.
+    # .add_field(sql="MAP(ARRAY[ROW(1, 2)], ARRAY[CAST(null AS VARCHAR)])", python={(1, 2): None})
```

Comment:
follow-up issue for the TODO?

Reply:
https://github.com/trinodb/trino-python-client/issues/442